proofs/alf/cvc5/programs/Strings.smt3

```diff
@@ -587,3 +587,147 @@
+;   in SMT", CAV 2019. When applicable, we also rely on calls that reason that string
+;   length is non-negative and basic properties of arithmetic.
+(program $str_arith_entail_is_approx_len ((s String) (t String) (r String) (n1 Int) (n2 Int) (n Int) (isUnder Bool))
+  (String Int Bool) Bool
+  (
+    (($str_arith_entail_is_approx_len (str.substr s n1 n2) n isUnder)
+        (let ((npm (+ n1 n2)))
+        (let ((lx (str.len s)))
+        (alf.ite (alf.is_eq n n2)
+          (alf.ite isUnder
+            (alf.and ($str_arith_entail_simple n1) ($str_arith_entail_simple_pred (>= lx npm)))
+            ($str_arith_entail_simple n2))
+        (alf.ite (alf.is_eq n (str.len s))
+          (alf.not isUnder)
+        (alf.ite (alf.is_eq n (- (str.len s) n1))
+          (alf.ite isUnder
+            (alf.and ($str_arith_entail_simple n1) ($str_arith_entail_simple_pred (>= npm lx)))
+            ($str_arith_entail_simple_pred (>= lx n1)))
+          false))))))
+    (($str_arith_entail_is_approx_len (str.replace s t r) n isUnder)
+        (let ((ls (str.len s)))
+        (let ((lt (str.len t)))
+        (let ((lr (str.len r)))
+        (alf.ite (alf.is_eq n ls)
+          (alf.ite isUnder
+            (alf.or ($str_arith_entail_simple_pred (>= lr lt)) ($str_arith_entail_simple_pred (>= lr ls)))
+            ($str_arith_entail_simple_pred (>= lt lr)))
+        (alf.ite (alf.is_eq n (+ ls lr))
+          (alf.not isUnder)
+        (alf.ite (alf.is_eq n (- ls lt))
+          isUnder
+          false)))))))
+    (($str_arith_entail_is_approx_len (str.from_int n1) n isUnder)
+        (alf.ite (alf.is_eq n n1)
+          (alf.and (alf.not isUnder) ($str_arith_entail_simple n1))
+        (alf.ite (alf.is_eq n (+ n1 1))
+          (alf.and (alf.not isUnder) ($str_arith_entail_simple_pred (> n1 0)))
+        (alf.ite (alf.is_eq n 1)
+          (alf.and isUnder ($str_arith_entail_simple n1))
+          false))))
+  )
+)
+
+; program: $str_arith_entail_is_approx_len
+; args:
+; - s String: The first string argument to str.indexof.
+; - t String: The second string argument to str.indexof.
+; - n Int: The third argument of str.indexof.
+; - m Int: An integer that may under (resp. over) approximate the given indexof term when b is true (resp. false).
+; - b Bool: Whether we are checking that m is an under (resp. over) approximation.
+; return: >
+;   true if we can show that (>= (str.indexof s t n) m) (or <= if b is false) holds using
+;   approximation cases as defined by Figure 2 of "High-Level Abstractions for Simplifying
+;   Extended String Constraints in SMT", CAV 2019. When applicable, we also rely on calls
+;   that reason that string length is non-negative and basic properties of arithmetic.
+(define $str_arith_entail_is_approx_indexof ((s String) (t String) (n Int) (m Int) (isUnder Bool))
+  (let ((ls (str.len s)))
+  (let ((lt (str.len t)))
+  (alf.ite (alf.is_eq m -1)
+    isUnder
+  (alf.ite (alf.is_eq m ls)
+    (alf.not isUnder)
```

Comment:
Note: it can be an under-approximation if `n` is the length of `s` and `t` is empty.

Reply:
Indeed there are more cases to consider, but this should be addressed in the solver, not the proof rule level.

We should consider making the approximations more precise and updating the proof rule in tandem with the solver.